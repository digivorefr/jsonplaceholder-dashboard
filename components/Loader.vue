<template>
  <div class="loader" />
</template>

<script lang="
    ts"
  >
import Vue from 'vue'

export default Vue.extend({
  name: 'ComponentLoader'
})
</script>
  </div>
</template>
