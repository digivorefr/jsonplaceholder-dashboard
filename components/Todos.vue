<template>
  <div :class="`user__todos${($fetchState.error !== null) ? ' error' : ''}`">
    <h2>Todos</h2>
    <h3 v-if="$fetchState.error !== null">
      An error occured.
    </h3>
    <Loader v-else-if="$fetchState.pending === true" />
    <h3 v-else-if="todos.length === 0">
      No todos
    </h3>
    <div
      v-else
      class="user__todos__wrapper"
    >
      <div class="user__todos__grid">
        <label
          v-for="(todo, index) in todos"
          :key="todo.id"
          :class="`user__todo${todo.completed ? ' completed' : ''}`"
          :for="`todo${todo.id}`"
        >
          <input
            :id="`todo${todo.id}`"
            type="checkbox"
            :checked="todo.completed"
            class="user__todo__checkbox"
            @change="() => onCheckboxChange(index)"
          >
          <div :class="`icon icon${todo.completed ? '--validate--success' : '--circle--light'} user__todo__icon`" />
          <p class="user__todo__content">{{ todo.title }}</p>
        </label>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import Loader from '~/components/Loader.vue'

export type Todo = {
  userId: number;
  id: number;
  title: string;
  completed: boolean;
}

export default Vue.extend({
  name: 'UserTodos',
  components: { Loader },
  props: {
    id: {
      type: Number,
      required: true
    }
  },
  data() {
    return {
      todos: [] as Todo[]
    }
  },
  async fetch() {
    this.todos = await this.getTodosByUserId()
    this.sortTodos()
  },
  methods: {
    async getTodosByUserId(): Promise<Todo[]> {
      return await this.$axios.$get(`/todos?userId=${this.id}`)
    },
    onCheckboxChange(index: number) {
      this.todos[index].completed = !this.todos[index].completed
      this.sortTodos()
    },
    sortTodos() {
      this.todos = this.todos
        .filter(todo => !todo.completed)
        .concat(
          this.todos.filter(todo => todo.completed)
        )
    }
  }
})
</script>
