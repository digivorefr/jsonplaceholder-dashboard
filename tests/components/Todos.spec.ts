import { mount, Wrapper } from '@vue/test-utils'
import Todos from '@/components/Todos.vue'
import Loader from '@/components/Loader.vue'

const todosMock = [
  {
    userId: 1,
    id: 1,
    title: 'delectus aut autem',
    completed: false
  },
  {
    userId: 1,
    id: 2,
    title: 'quis ut nam facilis et officia qui',
    completed: false
  }
]

describe('Todos', () => {
  test('renders correctly', () => {
    const wrapper = mount(Todos, {
      mocks: {
        $fetchState: { pending: true, error: true, timestamp: Date.now() }
      }
    })
    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.exists()).toBe(true)
    wrapper.destroy()
  })
  test('show error if fetch fails', () => {
    const wrapper = mount(Todos, {
      mocks: {
        $fetchState: { pending: null, error: true, timestamp: Date.now() }
      }
    })
    expect(wrapper.html()).toContain('An error occured.')
    wrapper.destroy()
  })
  test('show loader while fetching', () => {
    const wrapper = mount(Todos, {
      mocks: {
        $fetchState: { pending: true, error: null, timestamp: Date.now() }
      }
    })
    expect(wrapper.findComponent(Loader))
    wrapper.destroy()
  })
  test('show no todos text', () => {
    const wrapper = mount(Todos, {
      data() {
        return { todos: [] }
      },
      mocks: {
        $fetchState: { pending: null, error: null, timestamp: Date.now() }
      }
    })
    expect(wrapper.html()).toContain('No todos')
    wrapper.destroy()
  })
  test('fetch and display todos', async () => {
    const wrapper: Wrapper<Vue & { [key: string]: any }> = mount(Todos, {
      mocks: {
        $fetchState: { pending: null, error: null, timestamp: Date.now() },
        $axios: { $get: () => todosMock }
      }
    })
    wrapper.vm.$data.todos = await wrapper.vm.getTodosByUserId()
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.$data.todos[0].id).toBe(1)
    expect(wrapper.findAll('.user__todo').length).toBe(2)
    expect(wrapper.element).toMatchSnapshot()
    wrapper.destroy()
  })
  test('Toggle todos', async () => {
    const wrapper = mount(Todos, {
      data() {
        return { todos: todosMock }
      },
      mocks: {
        $fetchState: { pending: null, error: null, timestamp: Date.now() }
      }
    })
    const todo = wrapper.find('#todo1')
    const label = wrapper.find('label[for="todo1"]')
    await todo.setChecked()

    expect(label.classes().includes('completed')).toBeTruthy()
    expect(wrapper.vm.$data.todos[1].id).toBe(1)

    await todo.setChecked(false)
    expect(label.classes().includes('completed')).toBeFalsy()
    expect(wrapper.vm.$data.todos[1].id).toBe(1)

    wrapper.destroy()
  })
})
