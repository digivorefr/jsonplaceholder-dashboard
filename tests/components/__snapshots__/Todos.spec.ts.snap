// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos fetch and display todos 1`] = `
<div
  class="user__todos"
>
  <h2>
    Todos
  </h2>
   
  <div
    class="user__todos__wrapper"
  >
    <div
      class="user__todos__grid"
    >
      <label
        class="user__todo"
        for="todo1"
      >
        <input
          class="user__todo__checkbox"
          id="todo1"
          type="checkbox"
        />
         
        <div
          class="icon icon--circle--light user__todo__icon"
        />
         
        <p
          class="user__todo__content"
        >
          delectus aut autem
        </p>
      </label>
      <label
        class="user__todo"
        for="todo2"
      >
        <input
          class="user__todo__checkbox"
          id="todo2"
          type="checkbox"
        />
         
        <div
          class="icon icon--circle--light user__todo__icon"
        />
         
        <p
          class="user__todo__content"
        >
          quis ut nam facilis et officia qui
        </p>
      </label>
    </div>
  </div>
</div>
`;
